TextIndicator needs to update with more complicated window changes.

megangardner opened the pull request:
#### 80d1a66176baafcd33e459d95f1c63327b4942f6
<pre>
TextIndicator needs to update with more complicated window changes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=292775">https://bugs.webkit.org/show_bug.cgi?id=292775</a>
<a href="https://rdar.apple.com/151011352">rdar://151011352</a>

Reviewed by Wenson Hsieh.

In preparation for having the text indicator update with more complicated
window changes, we need to move it into the main window. This also
unifies the path for text indicator to be the same on iOS and macOS.
Behavior should be mostly unchanged, but you will note that the bounce
will no longer travel outside the main widow, which will only come up
when the indicator is close to the edge. This is a tradeoff we are
willing to make so that the indicator can be more closely tied to
more complicated window updates, and will not be out of sync during
those transitions.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
(WebKit::WebViewImpl::renewGState):
(WebKit::WebViewImpl::setTextIndicator):
(WebKit::WebViewImpl::updateTextIndicator):
(WebKit::WebViewImpl::clearTextIndicatorWithAnimation):
(WebKit::WebViewImpl::setTextIndicatorAnimationProgress):
(WebKit::WebViewImpl::teardownTextIndicatorLayer):
(WebKit::WebViewImpl::startTextIndicatoreFadeOut):

Canonical link: <a href="https://commits.webkit.org/294860@main">https://commits.webkit.org/294860@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15a6d97ba22d7283bf6aff10589e88149853dfbe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22497 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12817 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107990 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53465 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30999 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78188 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35151 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105830 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17672 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58520 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17518 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52822 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87326 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10899 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110365 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29961 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22077 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87173 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30325 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88992 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86788 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22182 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31623 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9346 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24223 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29888 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35210 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29696 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33023 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31258 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -470,6 +471,8 @@ class WebViewImpl final : public CanMakeWeakPtr<WebViewImpl>, public CanMakeChec
     void clearTextIndicatorWithAnimation(WebCore::TextIndicatorDismissalAnimation);
     void setTextIndicatorAnimationProgress(float);
+    void teardownTextIndicatorLayer();
+    void startTextIndicatoreFadeOut();
```
```suggestion
    void startTextIndicatorFadeOut();
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3460,35 +3459,72 @@ static String commandNameForSelector(SEL selector)
+
+void WebViewImpl::clearTextIndicatorWithAnimation(WebCore::TextIndicatorDismissalAnimation animation)
+{
+    RefPtr<WebCore::TextIndicator> textIndicator = WTFMove(m_textIndicator);
```
```suggestion
    RefPtr textIndicator = m_textIndicator;
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3460,35 +3459,72 @@ static String commandNameForSelector(SEL selector)
+{
+    RefPtr<WebCore::TextIndicator> textIndicator = WTFMove(m_textIndicator);
+
+    if ([m_textIndicatorLayer isFadingOut])
+        return;
```
The early return can be moved to the top of the method.

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3460,35 +3459,72 @@ static String commandNameForSelector(SEL selector)
+
+void WebViewImpl::startTextIndicatoreFadeOut()
+{
+    [m_textIndicatorLayer setFadingOut:YES];
```
Hm... I wonder if `WebTextIndicatorLayer` should just internally update `-isFadingOut` (instead of having the call site manage it). May also want to consider either resetting `-isFadingOut` when the fade animation completes, or renaming it to something like `-isFadingOutOrHidden`

megangardner replied:
I think there's a lot of work to be done to clean this up now that we only have one code path, but I'm taking an iterative approach.

megangardner committed:
TextIndicator needs to update with more complicated window changes.
https://bugs.webkit.org/show_bug.cgi?id=292775
rdar://151011352

Reviewed by Wenson Hsieh.

In preparation for having the text indicator update with more complicated
window changes, we need to move it into the main window. This also
unifies the path for text indicator to be the same on iOS and macOS.
Behavior should be mostly unchanged, but you will note that the bounce
will no longer travel outside the main widow, which will only come up
when the indicator is close to the edge. This is a tradeoff we are
willing to make so that the indicator can be more closely tied to
more complicated window updates, and will not be out of sync during
those transitions.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
(WebKit::WebViewImpl::renewGState):
(WebKit::WebViewImpl::setTextIndicator):
(WebKit::WebViewImpl::updateTextIndicator):
(WebKit::WebViewImpl::clearTextIndicatorWithAnimation):
(WebKit::WebViewImpl::setTextIndicatorAnimationProgress):
(WebKit::WebViewImpl::teardownTextIndicatorLayer):
(WebKit::WebViewImpl::startTextIndicatoreFadeOut):

Canonical link: https://commits.webkit.org/294860@main
